src/context.rs

```diff
@@ -0,0 +1,178 @@
+use std::collections::HashMap;
+use std::any::Any;
+
+#[derive(Debug,Clone,PartialEq)]
+pub enum Variable {
+    String(String),
+    Number(f64),
+    Boolean(bool),
+    //other(Drop), coming soon(tm)
+}
+
+pub struct Context {
+    variables: Vec<HashMap<String, Variable>>,
+    current_scope: usize,
+}
+
+impl Context {
+    pub fn new() -> Context {
+        let variables = HashMap::new();
+        Context { variables: vec![variables], current_scope: 0 }
+    }
+
+    fn current_variables(&mut self) -> &mut HashMap<String, Variable> {
+        &mut self.variables[self.current_scope]
+    }
+
+    pub fn add<T: Any>(&mut self, key: &str, val: &T) {
+        let val_any = val as &Any;
+
+        if let Some(string) = val_any.downcast_ref::<&str>() {
+            self.current_variables().insert(key.to_string(), Variable::String(string.to_string()));
+        } else if let Some(number) = val_any.downcast_ref::<f64>() {
+            self.current_variables().insert(key.to_string(), Variable::Number(*number));
+        } else if let Some(boolean) = val_any.downcast_ref::<bool>() {
+            self.current_variables().insert(key.to_string(), Variable::Boolean(*boolean));
+        } else {
+            panic!("Tried to add unknown type to context");
+        }
+
+    }
+
+    pub fn lookup(&mut self, key: &str) -> Option<&Variable> {
+        for scope in self.variables.iter().rev() {
+            match scope.get(key) {
+                Some(val)   => return Some(val),
+                None        => continue,
+            }
+        }
+        None
+    }
+
+    pub fn push(&mut self) {
+        self.variables.push(HashMap::new());
+        self.current_scope += 1;
+    }
+
+    pub fn pop(&mut self) {
+        if self.current_scope >= 1 {
+            self.variables.pop();
+            self.current_scope -= 1;
+        } else {
+            panic!("tried to pop one too many scopes!");
```

Comment:
Would the pop call panic if current_scope < 1? If so, we could just let that happen rather than guard just to panic